File: basic/src/main/java/com/yuanshenbin/basic/state/StateAbstract.java
package com.yuanshenbin.basic.state;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.LinearLayout;

/**
 * author : yuanshenbin
 * time   : 2017/12/7
 * desc   :
 */

public abstract class StateAbstract {
    protected Context mContext;
    protected LinearLayout root;

    @LayoutRes
    public abstract int onCreateView();


    public void onViewCreated(Context context, LinearLayout root) {
        this.mContext = context;
        this.root = root;
        if (root.getVisibility() == View.GONE) {
            root.setVisibility(View.VISIBLE);
        }
        removeViews(root);
        ViewStub viewStub = new ViewStub(context);
        viewStub.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
        viewStub.setLayoutResource(onCreateView());
        root.addView(viewStub);
        viewStub.inflate();
    }

    public void removeViews(LinearLayout root) {
        root.removeAllViews();
    }

} 

File: basic-kotlin/src/main/java/com/yuanshenbin/basic/state/StateAbstract.kt
package com.yuanshenbin.basic.state

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.view.ViewStub
import android.widget.LinearLayout
import androidx.annotation.LayoutRes

/**
 * author : yuanshenbin
 * time   : 2017/12/7
 * desc   :
 */
abstract class StateAbstract {
    protected var mContext: Context? = null
    protected var root: LinearLayout? = null

    @LayoutRes
    abstract fun onCreateView(): Int
    open  fun onViewCreated(context: Context?, root: LinearLayout) {
        mContext = context
        this.root = root
        if (root.visibility == View.GONE) {
            root.visibility = View.VISIBLE
        }

        removeViews(root)
        val viewStub = ViewStub(context)
        viewStub.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        viewStub.layoutResource = onCreateView()
        root.addView(viewStub)
        viewStub.inflate()
    }

    fun removeViews(root: LinearLayout) {
        root.removeAllViews()
    }
}